File: queues/finding_mk_avg.py
'''
LeetCode - problem 1825
'''

from sortedcontainers import SortedList

class MKAverage:

    def __init__(self, m: int, k: int):
        self.arr = SortedList()
        self.m = m
        self.k = k
        self.q = deque()
        self.total = None

    def addElement(self, num: int) -> None:
        self.q.append(num)
        m = self.m
        k = self.k
        if len(self.q) > m:
            val = self.q.popleft()            
            ind1 = self.arr.bisect_left(val)
            ind2 = self.arr.bisect_right(val)
            left, right, mid = False, False, False
            kth, mth = self.arr[k], self.arr[m - k - 1]
            if k <= ind1 < m - k or k <= ind2 - 1 < m - k or (ind1 <= k and  m - k <= ind2 - 1 < m):
                mid = True
            elif ind1 < k:
                left = True
            elif ind2 - 1 >= m - k:
                right = True            
            
            self.arr.remove(val)
            ind1 = self.arr.bisect_left(num)
            ind2 = self.arr.bisect_right(num)            
            self.arr.add(num)            
            
            if k <= ind1 < m - k or k <= ind2 < m - k or (ind1 <= k and  m - k <= ind2 < m):
                if mid:
                    self.total += num - val
                if left:
                    self.total += num - kth
                if right:
                    self.total += num - mth
            elif ind1 < k:
                if mid:
                    self.total += self.arr[k] - val
                if left:
                    pass
                if right:
                    self.total += self.arr[k] - mth
            elif ind2 >= m - k:
                if mid:
                    self.total += self.arr[m - k - 1] - val
                if left:
                    self.total += self.arr[m - k - 1] - kth
                if right:
                    pass
        else:
            self.arr.add(num)
            if self.total is None and len(self.arr) == self.m:
                self.total = sum(self.arr[k:m-k])
            

    def calculateMKAverage(self) -> int:
        if len(self.arr) < self.m:
            return -1
        
        return int(self.total / (self.m - 2 * self.k))

File: queues/finding_mk_avg.cpp
//LeetCode - problem 1825

class MKAverage {
public:
    multiset<int> left,mid,right;
    int m,k;
    long long s;
    vector<int> t;
    MKAverage(int m, int k) {
        this->m=m;
        this->k=k;
        s=0;
    }
    
    void addElement(int num) {
        t.push_back(num);
        if(t.size()<=m) {
            if(left.size()<k || num<*rbegin(left)) 
            {
                left.insert(num);
                if(left.size()>k) 
                {
                    mid.insert(*rbegin(left));
                    s+=*rbegin(left);
                    left.erase(prev(end(left)));
                    if(right.size()<k || *rbegin(mid)>*begin(right)) {
                        right.insert(*rbegin(mid));
                        s-=*rbegin(mid);
                        mid.erase(prev(end(mid)));
                        
                        if(right.size()>k) {
                            mid.insert(*begin(right));
                            s+=*begin(right);
                            right.erase(begin(right));
                        }
                    }
                    
                }
            } else {
                mid.insert(num);
                s+=num;
                if(right.size()<k||*rbegin(mid)>*begin(right)) {
                    right.insert(*rbegin(mid));
                    s-=*rbegin(mid);
                    mid.erase(prev(end(mid)));
                    if(right.size()>k) {
                        mid.insert(*begin(right));
                        s+=*begin(right);
                        right.erase(begin(right));
                    }
                }
            }
        }
        else {
                //delete val
                int val=t[t.size()-m-1];
                if(left.find(val)!=left.end()) {
                    left.erase(left.lower_bound(val));
                    left.insert(*begin(mid));
                    s-=*begin(mid);
                    mid.erase(begin(mid));
                } else if(mid.find(val)!=mid.end()) {
                    s-=val;
                    mid.erase(mid.lower_bound(val));
                } else {
                    right.erase(right.lower_bound(val));
                    right.insert(*rbegin(mid));
                    s-=*rbegin(mid);
                    mid.erase(prev(end(mid)));
                }
                //add num
                if(num<*rbegin(left)) {
                    left.insert(num);
                    mid.insert(*rbegin(left));
                    s+=*rbegin(left);
                    left.erase(prev(end(left)));
                } else {
                    mid.insert(num);
                    s+=num;
                    if(*rbegin(mid)>*begin(right)) {
                        right.insert(*rbegin(mid));
                        s-=*rbegin(mid);
                        mid.erase(prev(end(mid)));
                        mid.insert(*begin(right));
                        s+=*begin(right);
                        right.erase(begin(right));
                    }
                }
            }

    }
    
    int calculateMKAverage() {
    if(left.size()+right.size()+mid.size()<m) return -1;
    return s/(mid.size());

    }
};

